msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
         return iterator
 
 
-# This is the initial prototype interface for sim_ancestry. A bunch of
-# parameters have been left out in the interest of simplicity so that
-# we can get an initial version in as soon as possible.
-def sim_ancestry(
-    sample_size=None,
+def _parse_rate_map(rate_param, sequence_length, name):
+    """
+    Parse the specified input rate parameter value into a rate map.
+    """
+    # Note: in the future we might have another clause here where we
+    # allow for a different  map per population. This could be
+    # accepted as either a list of N rate maps, or a dictionary mapping
+    # population names to maps.
+    # See https://github.com/tskit-dev/msprime/issues/1095
+
+    msg_head = f"Error in parsing rate map for {name}: "
+    if isinstance(rate_param, intervals.RateMap):
+        rate_map = rate_param
+        if rate_map.sequence_length != sequence_length:
+            raise ValueError(msg_head + "sequence_length must match")
+    else:
+        rate_param = 0 if rate_param is None else float(rate_param)
+        rate_map = intervals.RateMap.uniform(sequence_length, rate_param)
+    return rate_map
+
+
+def _insert_integer_samples(n, ploidy, tables):
+    """
+    Insert n individuals with the specified ploidy into the tables.
+    """
+    # Doing this with numpy is slighty obscure, but it's *much* faster
+    # than simple loops. For 10^7 samples with ploidy=2 we go from
+    # several minutes to a few seconds.
+    ind_flags = np.zeros(n, dtype=np.uint32)
+    tables.individuals.set_columns(flags=ind_flags)
+    N = n * ploidy
+    node_individual = np.zeros(N, dtype=np.int32)
+    a = np.arange(n, dtype=node_individual.dtype)
+    for j in range(ploidy):
+        node_individual[j::ploidy] = a
+    tables.nodes.set_columns(
+        flags=np.full(N, tskit.NODE_IS_SAMPLE, dtype=np.uint32),
+        time=np.zeros(N),
+        population=np.zeros(N, dtype=np.int32),
+        individual=node_individual,
+    )
+
+
+def _parse_samples(samples, ploidy, tables):
+    """
+    Parse the specified "samples" value and insert them into the specified
+    tables. How the samples argument is interpreted depends on the state
+    of the tables.
+    """
+    if isinstance(samples, collections.abc.Iterable):
+        # We check if the samples can be interpreted as a list first because
+        # we may want to extend the semantics here later to include specifying
+        # a set of individuals from a pedigree defined in the initial state.
+        # The most natural way to do this would be a numpy array of individual
+        # IDs. By checking the iterable condition first, we make sure that we're
+        # not interpreting input numpy arrays of length 1 as a numeric argument.
+        # Interpret as a list of Sample objects.
```

Comment:
maybe the following should be done with numpy, also, e.g.
```
pops = np.array([s.population for s in samples])
node_pops = np.repeat(pops, ploidy)
node_inds = np.repeat(range(n), ploidy)
... set_columns...
```

Reply:
Good call. Opened #1211 to track.